File: gran_seoul.rb
require 'selenium-webdriver'
require 'faraday'
require 'faraday-cookie_jar'
require 'nokogiri'
require 'json'


KAKAO_HOST_URL = 'https://parking.kakao.com/admin'
AREA_NAME = '알파리움'


HOST_URL = 'http://npgsgranparkingroot.iptime.org'.freeze
LOGIN_URL = '/j_spring_security_check'.freeze
SEARCH_CAR_NUMBER_URL = '/ezTicket/carSearch'.freeze
SEARCH_TICKET_URL = '/ezTicket/carTicket'.freeze
GET_PARK_CODE_URL = '/ezTicket/inputDiscountValue'.freeze
ADD_COUPON_URL = '/ezTicket/valiDiscountCode.json'.freeze
SUBMIT_COUPON_URL = '/ezTicket/inputDiscountValueRegister'.freeze
GRAN_SEOUL_AREA_ID = 58

COUPONS = [{ id: 60, name: '30분쿠폰', value: 30 },
           { id: 61, name: '1시간쿠폰', value: 60 },
           { id: 62, name: '2시간쿠폰', value: 120 },
           { id: 63, name: '3시간쿠폰', value: 180 },
           { id: 64, name: '4시간쿠폰', value: 240 }].freeze


@conn = Faraday.new(url: HOST_URL) do |faraday|
 faraday.request :url_encoded
 faraday.response :logger
 faraday.use :cookie_jar
 faraday.adapter Faraday.default_adapter
end




def initialize(a, b, c)
  # raise 'Reservation is nil' if reservation.nil?
  # raise 'Reservation is not from GranSeoul' unless reservation.area.id.eql?(GRAN_SEOUL_AREA_ID)
  # raise 'Product is nil' if reservation.oper_table_product.nil?

  @real_ti = nil
  @park_code = nil
  @click_count = 0
  @order = { discount_value: '',
             discount_name: [],
             discount_code: [],
             discount_desc: [] }
  @car_number = b.gsub(/[\s+-]/, '')
  @ti = a
  @duration = c
  # init_connection
end

def login?
  response = @conn.post LOGIN_URL, j_username: 'parkhere', j_password: '1234'
  !response.env.response_headers[:location].include?('fail')
end

def find_car_number?
  is_find = false
  response = @conn.post SEARCH_CAR_NUMBER_URL, car_no: @car_number.to_s[-4, 4]
  if response.status.eql?(200)
    Nokogiri::HTML(response.body).css('table > tbody > tr').each do |tr|
      car_number = tr.css('a:first').text unless tr.css('a:first').empty?
      next unless @car_number.eql?(car_number)
      next if tr.css('input:first').empty?
      real_ti = Time.zone.parse(tr.css('input:first').attr('value'))
      next unless real_ti.year.eql?(@ti.year) && real_ti.month.eql?(@ti.month) && real_ti.day.eql?(@ti.day)
      @real_ti = real_ti
      set_duration
      get_park_code
      is_find = true
      break
    end
  end
  is_find
end

def process?
  make_order
  submit_coupon
  is_process = submit_coupon_success?
  logger.debug 'Could not find submit result' unless is_process
  is_process
end

def self.area_id
  GRAN_SEOUL_AREA_ID
end

def get_park_code
  response = @conn.post GET_PARK_CODE_URL, car_no: @car_number, car__totaldate: @real_ti.strftime('%Y%m%d%H%M%S')
  # raise RuntimeError unless response.status.eql?(200)
  park_code_list = Nokogiri::HTML(response.body).xpath('//input[@id="park_code"]')
  @park_code = park_code_list.first.attr('value') unless park_code_list.empty?
end

def make_order
  coupon_times = COUPONS.map { |elem| elem[:value] }
  duration = @duration
  while duration > 0 && !coupon_times.empty?
    if duration >= coupon_times[-1]
      coupon = COUPONS.find { |elem| elem[:value] == coupon_times[-1] }
      add_coupon(coupon)
      duration -= coupon[:value]
    else
      coupon_times.pop
    end
  end

  return unless duration > 0
  coupon = COUPONS.find { |elem| elem[:value] == 30 }
  add_coupon(coupon)
end

def add_coupon(coupon)
  response = @conn.post ADD_COUPON_URL, discount_code: coupon[:id], car_no: @car_number, clickCount: @click_count
  json_body = JSON.parse(response.body)
  @order[:discount_value] = @order[:discount_value].to_i + json_body['dcValue'].to_i
  @order[:discount_name].push(coupon[:name])
  @order[:discount_code].push(coupon[:id])
  @order[:discount_desc].push(json_body['dcDesc'])

  @click_count += 1
end

def submit_coupon
  params = { group_name: '모바일 할인',
             department_name: '파크히어',
             car_no: @car_number,
             park_code: @park_code,
             car_in_date: "#{@real_ti.strftime('%Y-%m-%d')} 00:00:00.0",
             car_in_time: @real_ti.strftime('%H:%M'),
             car__totaldate: @real_ti.strftime('%Y%m%d%H%M%S'),
             discount_name: @order[:discount_name].join(','),
             discount_code: @order[:discount_code].join(','),
             discount_value: @order[:discount_value],
             discount_desc: @order[:discount_desc].join(',') }
  logger.debug params
  @conn.post SUBMIT_COUPON_URL, params
end

def submit_coupon_success?
  response = @conn.post SEARCH_TICKET_URL
  return false unless response.status.eql?(200)

  Nokogiri::HTML(response.body).css('table > tbody > tr').each do |tr|
    td_texts = tr.css('td').map(&:text)
    return true if td_texts.include?(@car_number) && td_texts.include?(@duration.to_s)
  end
  false
end

def set_duration
  to = @real_ti + 60 * @duration
  return if to.day == @real_ti.day
  @duration -= to.hour * 60
  @duration -= (to.min / 30) * 30
end





GranSeoul = GranSeoul.new('2018-02-13 14:10:22', '27구4146', 1080)
if GranSeoul.login?
  puts 'login'
	if GranSeoul.find_car_number?
    puts 'find car number!'
	GranSeoul.process?
  else
    puts 'cant find car number :('
  end
end

Selenium::WebDriver::Chrome.driver_path = ('/Users/gilsanghyeog/Documents/chromedriver')
driver = Selenium::WebDriver.for :chrome
driver.get(HOST_URL + '/login')
driver.find_element(name: 'manager[login]').send_keys('san.kill')
driver.find_element(name: 'manager[password]').send_keys('!@dl926516')
driver.find_element(name: 'commit').click()
driver.get(HOST_URL + '/picks?q[parking_lot_name_cont]=' + AREA_NAME + '&q[state_eq]=4&order=id_desc')
driver.page_source

File: gran_seoul.py
#-*- coding: utf-8 -*-

import requests
from bs4 import BeautifulSoup
import sys
import time
import datetime
import json
from bot_interface import BotInterface

from importlib import reload



#웹할인 페이지 계정
LOGIN_INFO = {
    'j_username': 'kakao t',
    'j_password': '123456'
}
#웹할인 페이지
PARK_HOST_URL = 'http://npgsgranparkingroot0.iptime.org'
LOGIN_URL = '/j_spring_security_check'
SEARCH_CAR_NUMBER_URL = '/ezTicket/carSearch'
SEARCH_TICKET_URL = '/ezTicket/carTicket'
GET_PARK_CODE_URL = '/ezTicket/inputDiscountValue'
ADD_COUPON_URL = '/ezTicket/valiDiscountCode.json'
SUBMIT_COUPON_URL = '/ezTicket/inputDiscountValueRegister'

COUPONS = [{ 'id': 60, 'name': '30분쿠폰', 'value': 30 },
           { 'id': 61, 'name': '1시간쿠폰', 'value': 60 },
           { 'id': 62, 'name': '2시간쿠폰', 'value': 120 },
           { 'id': 63, 'name': '3시간쿠폰', 'value': 180 },
           { 'id': 64, 'name': '4시간쿠폰', 'value': 240 }]
AREA_ID = '1804'

class GranSeoulBot(BotInterface):
    def __init__(self, reservation):
        self.real_ti = ''
        self.park_code = ''
        self.click_count = 0
        self.order = { 'discount_value': 0,
                         'discount_name': [],
                         'discount_code': [],
                         'discount_desc': [] }
        self.k_car_num = reservation['k_car_num']
        self.ti = reservation['ti']
        self.entry_date = reservation['entry_date']
        self.duration = reservation['duration']

        self.s = requests.Session()

    def login(self):
        try:
            login_req = self.s.post(PARK_HOST_URL + LOGIN_URL, data=LOGIN_INFO)
            return True
        except requests.exceptions.ConnectionError:
            print('connection error')
            return False

    def find_car_number(self):
        flag = False
        find_req = self.s.post(PARK_HOST_URL + SEARCH_CAR_NUMBER_URL, data={'car_no': self.k_car_num[-4:]})
        if find_req.status_code == 200:
            soup = BeautifulSoup(find_req.content, 'html.parser')
            for tr in soup.select('table tbody tr'):
                if self.k_car_num != tr.select('a')[0].text: continue
                self.real_ti = tr.select('input')[0].get('value')
                self.real_ti = datetime.datetime.strptime(str(self.real_ti), "%Y%m%d%H%M%S")
                if self.real_ti.strftime("%Y-%m-%d") != self.entry_date: continue
                self.set_duration()
                self.get_park_code()
                flag = True
                break
        return flag

    def process(self):
        flag = True
        self.make_order()
        self.submit_coupon()
        return flag

    def make_order(self):
        coupon_times = list(map(lambda i: i['value'], COUPONS))
        duration = self.duration
        while duration > 0 and not not coupon_times:
            if duration >= coupon_times[-1]:
              coupon = list(filter(lambda i: i['value'] == coupon_times[-1], COUPONS))[0]
              self.add_coupon(coupon)
              duration -= coupon['value']
            else:
              coupon_times.pop()
        if not duration > 0: return True
        coupon = list(filter(lambda i: i['value'] == 30, COUPONS))[0]
        self.add_coupon(coupon)
        return True

    def add_coupon(self, coupon):
        response = self.s.post(PARK_HOST_URL + ADD_COUPON_URL, data={'discount_code': coupon['id'], 'car_no': self.k_car_num, 'clickCount': self.click_count})
        json_body = json.loads(response.content)
        self.order['discount_value'] = self.order['discount_value'] + json_body['dcValue']
        self.order['discount_name'].append(coupon['name'])
        self.order['discount_code'].append(coupon['id'])
        self.order['discount_desc'].append(json_body['dcDesc'])
        self.click_count += 1
        return True

    def set_duration(self):
        to = self.real_ti + datetime.timedelta(minutes = self.duration)
        if to.day == self.real_ti.day: return True
        self.duration -= to.hour * 60
        self.duration -= (to.minute / 30) * 30

    def get_park_code(self):
        get_park_req = self.s.post(PARK_HOST_URL + GET_PARK_CODE_URL, data={'car_no': self.k_car_num, 'car__totaldate': self.real_ti.strftime('%Y%m%d%H%M%S')})
        soup = BeautifulSoup(get_park_req.content, 'html.parser')
        self.park_code = soup.find(id='park_code').get('value')

    def submit_coupon(self):
        params = { 'group_name': '모바일 할인',
                 'department_name': '카카오',
                 'car_no': self.k_car_num,
                 'park_code': self.park_code,
                 'car_in_date': self.real_ti.strftime('%Y-%m-%d') + ' 00:00:00.0',
                 'car_in_time': self.real_ti.strftime('%H:%M'),
                 'car__totaldate': self.real_ti.strftime('%Y%m%d%H%M%S'),
                 'discount_name': ','.join(str(v) for v in self.order['discount_name']),
                 'discount_code': ','.join(str(v) for v in self.order['discount_code']),
                 'discount_value': self.order['discount_value'],
                 'discount_desc': ','.join(str(v) for v in self.order['discount_desc']) }
        self.s.post(PARK_HOST_URL + SUBMIT_COUPON_URL, data=params)

    def submit_coupon_success(self):
        response = self.s.post(PARK_HOST_URL + SEARCH_TICKET_URL)
        if not response.status_code == 200: return False
        soup = BeautifulSoup(response.content, 'html.parser')
        for tr in soup.select('table tbody tr'):
            tr = ','.join(str(v) for v in tr)
            if str(self.duration) in tr and self.k_car_num in tr: return True
        return False

    @staticmethod
    def area_id():
        return AREA_ID
